File: services/stream-processor/window.go
package main

import (
	"math"
	"sync"
	"time"
)

type sample struct {
	at    time.Time
	value float64
}

// windowStats summarizes a single (device, metric) series over one window
// duration: moving average, moving standard deviation, min, max, and rate of
// change (last-first value over the window's elapsed time) — the same shape
// covers "vibration trend" or "energy consumption rate" from the spec; those
// are just rate_of_change applied to the vibration/power metric respectively,
// not separately named computations.
type windowStats struct {
	Count        int
	MovingAvg    float64
	MovingStdDev float64
	Min          float64
	Max          float64
	RateOfChange float64
}

// seriesBuffer is a per-(device_id,metric) ring of recent samples, trimmed to
// the longest configured window on every insert so memory stays bounded
// regardless of how long the process runs.
type seriesBuffer struct {
	mu        sync.Mutex
	samples   []sample
	maxWindow time.Duration
}

func newSeriesBuffer(maxWindow time.Duration) *seriesBuffer {
	return &seriesBuffer{maxWindow: maxWindow}
}

// add inserts by event timestamp, not arrival order. Telemetry can arrive
// out of order (network delay, MQTT/Kafka redelivery — the simulator's
// OUT_OF_ORDER_RATE exercises this deliberately), and a pre-GitHub audit
// found this method simply appending regardless: with an out-of-order
// arrival, s.samples stops being sorted by .at, which corrupts two things
// that assume it is — the trim loop below (it only scans from the front,
// so a stale sample stuck behind a later-timestamped-but-earlier-arrived
// one never gets dropped) and computeStats's rate-of-change, which reads
// samples[0]/samples[n-1] as "chronologically first/last in the window"
// (an out-of-order append could make the "elapsed" denominator negative,
// or simply compute a rate between the wrong two readings).
func (s *seriesBuffer) add(at time.Time, value float64) {
	s.mu.Lock()
	defer s.mu.Unlock()

	i := len(s.samples)
	for i > 0 && s.samples[i-1].at.After(at) {
		i--
	}
	s.samples = append(s.samples, sample{})
	copy(s.samples[i+1:], s.samples[i:])
	s.samples[i] = sample{at: at, value: value}

	// Trim against the newest timestamp actually in the buffer (not this
	// call's `at`, which — for an out-of-order arrival — could be the
	// oldest thing in it and would under-trim).
	cutoff := s.samples[len(s.samples)-1].at.Add(-s.maxWindow)
	j := 0
	for j < len(s.samples) && s.samples[j].at.Before(cutoff) {
		j++
	}
	if j > 0 {
		s.samples = s.samples[j:]
	}
}

// statsFor computes windowStats over the trailing `window` duration, as of
// `now`. Returns ok=false if there are no samples in that window.
func (s *seriesBuffer) statsFor(now time.Time, window time.Duration) (windowStats, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()

	cutoff := now.Add(-window)
	var inWindow []sample
	for _, sm := range s.samples {
		if !sm.at.Before(cutoff) {
			inWindow = append(inWindow, sm)
		}
	}
	if len(inWindow) == 0 {
		return windowStats{}, false
	}
	return computeStats(inWindow), true
}

func computeStats(samples []sample) windowStats {
	n := len(samples)
	sum, min, max := 0.0, samples[0].value, samples[0].value
	for _, sm := range samples {
		sum += sm.value
		if sm.value < min {
			min = sm.value
		}
		if sm.value > max {
			max = sm.value
		}
	}
	avg := sum / float64(n)

	variance := 0.0
	for _, sm := range samples {
		d := sm.value - avg
		variance += d * d
	}
	variance /= float64(n)
	stddev := math.Sqrt(variance)

	rateOfChange := 0.0
	if n > 1 {
		elapsed := samples[n-1].at.Sub(samples[0].at).Seconds()
		if elapsed > 0 {
			rateOfChange = (samples[n-1].value - samples[0].value) / elapsed
		}
	}

	return windowStats{
		Count:        n,
		MovingAvg:    avg,
		MovingStdDev: stddev,
		Min:          min,
		Max:          max,
		RateOfChange: rateOfChange,
	}
}

File: services/stream-processor/window.py
"""Rolling per-series windowed statistics, the Python port of window.go.

Includes the same event-time-ordered insert this file's Go counterpart was
fixed to do after a pre-GitHub audit: telemetry can arrive out of order
(network delay, MQTT/Kafka redelivery — the simulator's OUT_OF_ORDER_RATE
exercises this deliberately), and a naive `samples.append()` would silently
break both the trim loop (which only scans from the front, so a stale
sample stuck behind a later-timestamped-but-earlier-arrived one never gets
dropped) and rate-of-change (which reads samples[0]/samples[-1] as
"chronologically first/last" — an out-of-order append could compute a
negative elapsed time, or simply the wrong two readings). See
tests/test_window.py for the regression tests ported 1:1 from window_test.go.
"""

from __future__ import annotations

import math
import threading
from dataclasses import dataclass
from datetime import datetime, timedelta


@dataclass
class Sample:
    at: datetime
    value: float


@dataclass
class WindowStats:
    """Summarizes a single (device, metric) series over one window
    duration: moving average, moving standard deviation, min, max, and rate
    of change (last-first value over the window's elapsed time) — the same
    shape covers "vibration trend" or "energy consumption rate" from the
    spec; those are just rate_of_change applied to the vibration/power
    metric respectively, not separately named computations."""

    count: int = 0
    moving_avg: float = 0.0
    moving_stddev: float = 0.0
    min: float = 0.0
    max: float = 0.0
    rate_of_change: float = 0.0


class SeriesBuffer:
    """A per-(device_id,metric) ordered list of recent samples, trimmed to
    the longest configured window on every insert so memory stays bounded
    regardless of how long the process runs."""

    def __init__(self, max_window_seconds: float):
        self._lock = threading.Lock()
        self._samples: list[Sample] = []
        self._max_window = timedelta(seconds=max_window_seconds)

    def add(self, at: datetime, value: float) -> None:
        """Inserts by event timestamp, not arrival order — see this
        module's docstring for why that matters."""
        with self._lock:
            i = len(self._samples)
            while i > 0 and self._samples[i - 1].at > at:
                i -= 1
            self._samples.insert(i, Sample(at=at, value=value))

            # Trim against the newest timestamp actually in the buffer (not
            # this call's `at`, which — for an out-of-order arrival — could
            # be the oldest thing in it and would under-trim).
            cutoff = self._samples[-1].at - self._max_window
            j = 0
            while j < len(self._samples) and self._samples[j].at < cutoff:
                j += 1
            if j > 0:
                self._samples = self._samples[j:]

    def stats_for(self, now: datetime, window_seconds: float) -> tuple[WindowStats, bool]:
        """Computes WindowStats over the trailing `window_seconds`, as of
        `now`. Returns ok=False if there are no samples in that window."""
        with self._lock:
            cutoff = now - timedelta(seconds=window_seconds)
            in_window = [s for s in self._samples if s.at >= cutoff]
        if not in_window:
            return WindowStats(), False
        return _compute_stats(in_window), True


def _compute_stats(samples: list[Sample]) -> WindowStats:
    n = len(samples)
    values = [s.value for s in samples]
    total = sum(values)
    lo, hi = min(values), max(values)
    avg = total / n

    variance = sum((v - avg) ** 2 for v in values) / n
    stddev = math.sqrt(variance)

    rate_of_change = 0.0
    if n > 1:
        elapsed = (samples[-1].at - samples[0].at).total_seconds()
        if elapsed > 0:
            rate_of_change = (samples[-1].value - samples[0].value) / elapsed

    return WindowStats(
        count=n,
        moving_avg=avg,
        moving_stddev=stddev,
        min=lo,
        max=hi,
        rate_of_change=rate_of_change,
    )
